src/kaleidoscope/device/dygma/raise/Hand.h

```diff
@@ -0,0 +1,117 @@
+/* -*- mode: c++ -*-
+ * kaleidoscope::device::dygma::Raise -- Kaleidoscope device plugin for Dygma Raise
+ * Copyright (C) 2017-2019  Keyboard.io, Inc
+ * Copyright (C) 2017-2019  Dygma Lab S.L.
+ *
+ * This program is free software: you can redistribute it and/or modify it under
+ * the terms of the GNU General Public License as published by the Free Software
+ * Foundation, version 3.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#ifdef ARDUINO_SAMD_RAISE
+
+#include <Arduino.h>
+#include "TWI.h"
+
+struct cRGB {
+  uint8_t r;
+  uint8_t g;
+  uint8_t b;
+};
+
+namespace kaleidoscope {
+namespace device {
+namespace dygma {
+namespace raise {
+
+#define LED_BANKS 9
+
+#define LEDS_PER_HAND 72
+#define LPH LEDS_PER_HAND
+#define LEDS_PER_BANK 8
+#define LED_BYTES_PER_BANK (sizeof(cRGB) * LEDS_PER_BANK)
+
+typedef union {
+  cRGB leds[LEDS_PER_HAND];
+  byte bytes[LED_BANKS][LED_BYTES_PER_BANK];
+} LEDData_t;
+
+// return what bank the led is in
+#define LED_TO_BANK(led) (led / LEDS_PER_BANK)
+
+// Same datastructure as on the other side
+typedef union {
+  struct {
+    uint8_t row: 2,
+            col: 3,
+            keyState: 1,
+            keyEventsWaiting: 1,
+            eventReported: 1;
+  };
+  uint8_t val;
+} key_t_raise;
+
+typedef union {
+  uint8_t rows[5];
+  uint64_t all;
+} keydata_t;
+
+class Hand {
+ public:
+  Hand(byte ad01) : ad01_(ad01), twi_(i2c_addr_base_ | ad01) {}
+
+  int readVersion();
+  int readSLEDVersion();
+  int readSLEDCurrent();
+  byte setSLEDCurrent(byte current);
+  int readJoint();
+  int readANSI_ISO();
+
+  byte setKeyscanInterval(byte delay);
+  int readKeyscanInterval();
+
+  byte setLEDSPIFrequency(byte frequency);
+  int readLEDSPIFrequency();
+
+  bool moreKeysWaiting();
+  void sendLEDData();
+  void sendLEDBank(uint8_t bank);
+  void setOneLEDTo(byte led, cRGB color);
+  void setAllLEDsTo(cRGB color);
+  keydata_t getKeyData();
+  bool readKeys();
+  uint8_t controllerAddress();
+  uint8_t crc_errors() {
+    return twi_.crc_errors();
+  }
+
+  LEDData_t led_data;
+  bool online = false;
+
+ private:
+  int ad01_;
+  TWI twi_;
+  keydata_t key_data_;
+  uint8_t next_led_bank_ = 0;
+
+  static constexpr uint8_t i2c_addr_base_ = 0x58;
```

Comment:
Aren't the dygma hands 0x50 and 0x51

Reply:
TWI uses `addr_base | ad01` for the address, and `ad01` is `0` or `1`... hmm. Might be a bug somewhere, 'cos that's 0x58 and 0x59 then.

FWIW, 0x50 and 0x51 are `LEFT_BOOT_ADDRESS` and `RIGHT_BOOT_ADDRESS` in Dygma's SideFlasher, while 0x58 is `SCANNER_I2C_ADDR_BASE`. I don't really know what these all mean.

Reply:
one set are for scanner mode. the other set are for bootloader made. They should probably get naming parity.